doc/performance/0002-http-vs-rcp-client-performance-analysis.md

```diff
@@ -0,0 +1,244 @@
+
+```text
+
+Params: experiment ID, output ID, and time range parameters
+Returns: the time graph model
+
+Test 3a
+Steps:
+1. Open the Theia Trace extension in a browser.
+2. Open a trace that has the Flame Chart output. 
+3. The flame chart should be opened and populated.
+4. Check the debug console and record the numbers.
+
+Test 3b
+Steps:
+1. Open the Theia Trace extension in a browser.
+2. Open a trace that has the Flame Chart output. 
+3. The flame chart should be opened and populated.
+4. Clear the console. This is to make sure that the right numbers are recorded.
+5. Perform a zoom with the right mouse button.
+6. Check the debug console and record the numbers.
+```
+
+The trace used for this benchmarking attempt can be found [here](https://github.com/eclipse-cdt-cloud/theia-trace-extension/pull/923).
+
+The zoom range used for test 3b is [1673902684.0-1673902684.5] (with the trace above).
+
+**Why do we need 2 tests?**
+
+In test 3a, the function `tspClient.fetchTimeGraphStates()` is executed, but there are some latency to reach the TSPClient because multiple requests are launched to load the chart. Hence, in test 3b, the measurement to reach the TSPClient is more accurate since there are no interfering processes. Also, in test 3b, there are around `116620 states` to be fetched from the server. This is to measure if the latency added by JSON-RPC correlates to the size of the HTTP response.
+
+### Sample test
+
+For example, to benchmark the function `getAvailableOutputs`, add the following logs:
+
+```typescript
+// In the Theia Trace Extension repo
+async getAvailableOutputs(experimentUUID: string): Promise<OutputDescriptor[] | undefined> {
+    // Save the timestamp when we initiate the communication to the TSPClient
+    const start = Date.now(); 
+    const outputsResponse = await this.fTspClient.experimentOutputs(experimentUUID);
+    // Save the timestamp when we receive the data back from the TSPClient
+    const end = Date.now();
+    // Output the timestamps to the debug console
+    console.log('Theia FE:: getAvailableOutputs:: start: ', start, ', end::', end);
+    if (outputsResponse && outputsResponse.isOk()) {
+        return outputsResponse.getModel();
+    }
+    return undefined;
+}
+
+// In the tsp-typescript-client repo
+public async experimentOutputs(
+        expUUID: string
+): Promise<TspClientResponse<OutputDescriptor[]>> {
+    // Save the timestamp when the TSPClient is reached
+    const start = Date.now();
+    const url = this.baseUrl + "/experiments/" + expUUID + "/outputs";
+    // Store the result in the variable to save a timestamp before returning
+    const result = await RestClient.get(url, undefined, array(OutputDescriptor));
+    // Save the timestamp when the HTTP request is finished
+    const end = Date.now();
+    // Output the timestamps to the debug console
+    console.log('TSPClient:: getAvailableOutputs:: client reached:', start, ', Promised resolved:', end);
+    return result;
+}
+```
+
```

Comment:
Could you please add the code for fetchXYTree and fetchTimeGraphStates so that it's easier for someone who wants to reproduce the tests.

Reply:
Ack